### DISCUSSION TITLE
Zip xcresult "folder" and prepare for exporting

### DISCUSSION THREAD

**Discussion by djups** [fastlane/fastlane#28791] [Show and tell]
Greetings! 
During scan lane, and UI test, we need to get and export to your cloud xcresult generated by xcode. 
Due fact, that xcresult is basically a folder, we have to zip it and upload to our services. Might be not important: next step we are adding link to xcresult.zip in our slack notifications. 

I was able to successfully zip xcresult with ruby zip gem. However, bash command "zip -r" are zipping wrong, and result zip file is corrupted.  

So, i propose to include in scan new parameter: zip_xcresult: Bool (false default), as a path we can use output_directory parameter. 

pseudocode:
  ```
require 'zip'

      scan(
       ....
      output_directory: "fastlane/testResultUI/#{scheme}",
     ....
    )
    
    xcresult_path = "../fastlane/testResultUI/#{scheme}.xcresult"
    zip_file_path = "../fastlane/testResultUI//#{scheme}.xcresult.zip"

  Zip::File.open(zip_file_path, Zip::File::CREATE) do |zipfile|
        Dir[File.join(xcresult_path, '**', '**')].each do |file|
          zipfile.add(file.sub(xcresult_path + '/', ''), file)
        end
      end
```

**Comment by bella964** [comment]
Hello,
Here’s a pseudocode implementation for your proposal. This assumes you are extending the scan tool in Fastlane to include the zip_xcresult parameter:     [meridian smart square](https://www.smart-squarehmh.com)
ruby
# Pseudocode for implementing zip_xcresult in scan
def scan(options)
  # Existing scan logic here...

  # Determine if zip_xcresult is enabled
  if options[:zip_xcresult]
    xcresult_path = File.join(options[:output_directory], "result.xcresult")
    
    # Check if xcresult exists
    if File.directory?(xcresult_path)
      zip_path = xcresult_path + ".zip"
      zip_xcresult_folder(xcresult_path, zip_path)
    else
      UI.error("xcresult folder not found at #{xcresult_path}")
    end
  end

  # Continue with existing scan process
end

def zip_xcresult_folder(source_path, destination_path)
  # Use Ruby's zip library to create a valid zip file
  require 'zip'
  Zip::File.open(destination_path, Zip::File::CREATE) do |zipfile|
    Dir[File.join(source_path, '**', '**')].each do |file|
      relative_path = file.sub(source_path + '/', '')
      zipfile.add(relative_path, file)
    end
  end

  UI.message("Successfully zipped xcresult to #{destination_path}")
rescue => e
  UI.error("Error while zipping xcresult: #{e.message}")
end
Explanation:
New Parameter: zip_xcresult is introduced with a default value of false.
Folder Path: The output_directory parameter is reused to locate the result.xcresult folder.
Ruby Zip Gem: The Ruby zip library is used to create a valid .zip file from the .xcresult folder.
Error Handling: Handles errors gracefully if the folder is missing or if zipping fails.
Slack Integration: If you have Slack integration in your workflow, the zipped file path can be attached to the Slack message.
Best Regards
bella964
